### DISCUSSION TITLE
RFC: Adding a subject browser to the classifier

### DISCUSSION THREAD

**Discussion by eatyourgreens** [zooniverse/front-end-monorepo#2166] [General]
Related to https://github.com/zooniverse/front-end-monorepo/discussions/2165

For Engaging Crowds, we'd like volunteers to be able to browse a subject set, and view subjects without actually classifying them. For example, while transcribing a page, I go back a couple of pages to look something up then return to my transcription and continue.

The Classifier isn't set up to browse subjects at the moment. In particular

- loading a subject into the viewer automatically starts a new classification. We want to be able to separate loading a subject from classifying, without breaking existing projects.
- the subject queue is a one-way stack that pops each classified subject off the stack in order to load the next. We want a mechanism that allows you to move back and forth through the current set, without disturbing the queue (I think.)

There's a service set up at https://subject-set-search-api.zooniverse.org which provides a JSON API for indexed subject sets, using [Datasette](https://datasette.io/). You can list subjects by set, select subjects by priority and generally treat a subject set as a browsable list of subjects. https://github.com/zooniverse/front-end-monorepo/pull/2148 has an example of integrating that API into the subject queue, in order to classify subjects ordered by metadata priority.

Maybe what's needed here is a new model, representing the Datasette API, but I'm not sure how that would integrate into the existing classifier.

**Comment by srallen** [comment]
I'm not sure why this feature would be in the classifier. This would be adding a lot of complexity to an already complex module and would make it so that it has multiple uses: classification and subject exploration. This feature sounds something that would better fit for a separate subject set/subject explorer UI perhaps something that could be launched from the classify page into a modal or inline widget below the classifier. Additionally, if you are classifying in order, the recents widget, already present on the classify page, or the yet to be built recents page, could serve as an exploratory tool as well with expanded design and features. 

What is the reason to have this be in the classifier itself?

**Comment by snblickhan** [comment]
To clarify why it's helpful to have this in the classifier, I think that it might help to use a different use case than the one Jim has shared here. For reference, here is the [original user stories doc](https://docs.google.com/document/d/1hQ6VlirvUBycve47fTf2Y7skGtcvY2zPhSQriJChGx8/edit)

A short user story for this feature might be:

"I want to choose a subject by the metadata available in the indexing tool. There's some metadata available, but not a lot, and I'm not actually that familiar with the documents in this project, so I just end up choosing one from the list at random. When it loads in the classifier, I'm not actually interested in this particular page, but because there isn't actually a specific one I'm looking for in the index, I just start hitting 'Next' on the Classify page until I see something I like and want to work on."

To be clear: the ability to look at other subjects _while classifying_ (e.g. the example Jim shared above) isn't in our MVP, so if it helps remove some complexity I'm happy to let that go (though I think it would be VERY cool, and aligns(?) with the CSSI slices goals, it isn't necessarily part of our research goals for Engaging Crowds). 

It may be that we need to create the same infrastructure change to support both of these behaviors anyway, but I think this helps to explain _why_ it's helpful to have this in the classifier. Ultimately, we're also answering a request that's been around at least since I started, which is what volunteers should do if they get to a subject they don't want to work on (without considering that subject 'already seen').

We have [design work for previous/next buttons on unavailable subjects](https://projects.invisionapp.com/d/main#/console/12924056/438011817/preview). @eatyourgreens do we have design for previous/next on available subjects? If so, can you post the link? If not, how will we incorporate these buttons into the classifier?

**Reply by eatyourgreens** [reply]
There are no designs for available subjects. We had a conversation, with Becky, about this during her last week (it might have been on her last day 😳 .) I'm not sure if we recorded that anywhere. It might only be in Slack, somewhere. I remember her saying that on an available subject, the only way to move on is to press Done and move to the next subject in the set. Then we discussed how that wasn't ideal if you're looking for a subject that you want to classify.

**Comment by snblickhan** [comment]
Follow-up user story & question:

User story: I am a volunteer, and I select a subject (Subject A) from the Index on a project. It takes me to Subject A in the classifier, but it isn't what I was expecting, and I don't actually want to work on it. I return to the Indexing tool to choose another subject. 

Question: if, in the scenario above, I looked at Subject A and then left the classifier (without interacting), would this action block me from working on Subject A in the future/cause Subject A to be set as 'Already Seen'? What action ends up moving a subject to 'Already Seen' -- starting a classification, or submitting one? If these view/abandon actions _don't_ prevent me from accessing Subject A later on, do we actually need to rethink when a classification starts? 

Hoping that this helps to clarify the 'what' and 'why' questions a bit further.

**Reply by srallen** [reply]
The `already_seen` property on subjects is something that is returned to us from Panoptes on the subject resource when we have an authenticated user. It only gets set if a previous classification has been submitted. We're missing the feature to track subjects already seen for anonymous (aka unauthenticated) users and authenticated users who are classifying on a workflow that has a small amount of subjects (so you get repeat subjects before the API has a chance to mark it as `'already_seen'` and what I believe is what #2111 is about). This session tracking would be part of the classification submission, not subject, workflow, and classification initialization.

What it sounds like is that we have two different features being conflated. One is a UI element to do what many volunteers on the Zooniverse already do, which is decide they don't want to work on the subject loaded and load a new one. Right now volunteers accomplish this by refreshing the page to get something new. What we could add is a button to show a new subject that could be used by any project. For projects that have the subject set ordered, it'll be the next subject in order. If some workflow tasks have been completed, then clicking this button would warn that a classification is in progress with a "yes show a new subject" and "no, I'd like to finish this one" as options. 

The other feature is to have a subject explorer. For comparison, the 'Slices' viewer for CSSI is to classify a subject in context of its 'surrounding' subjects and the subject viewer would have thumbnails that could be opened up for larger view of non-classifiable subjects that give context for the subject being actively classified. I'm not sure if this fits in with multi-image subjects like you often have with transcription. However, some work had begun with Designator to support this. A broader subject explorer I would still argue should be a separate feature apart from the classifier, but could be present on or launched from the classify page.

Lastly, we could have UI buttons on the classify page to open the selector modals to select a new workflow, subject set, or subject, and pass callback functions from the project app to manage this. We already do this with the collections modal.

**Comment by srallen** [comment]
Instead of next/previous buttons, could we just keep track of which page and parameters of the subject selection modal you last left off on and have a generic 'select a new subject' button that reopens where you left off?

**Reply by eatyourgreens** [reply]
That sounds like Becky's first iteration of the modal, in https://github.com/zooniverse/front-end-monorepo/issues/2027. It only has two buttons/links: Go to the home page, or choose a new subject.

She replaced that with the current version (https://github.com/zooniverse/front-end-monorepo/issues/2027#issuecomment-776219119) because we do want you to be able to browse through the subjects in the set, and see them, while you're deciding what to classify.

**Reply by srallen** [reply]
> because we do want you to be able to browse through the subjects in the set, and see them, while you're deciding what to classify.

Why though? I'm reading through the user stories, and I see this one as part of "Later", which I think maybe is what is driving this?

> As a user I want to click into subjects I’ve already classified and see a read-only classify interface (greyed out and unclickable).

This still doesn't explain to me why it has to be in the classify interface? Is it just to see the subject image larger before choosing to work on it? If so, we can do something that we've intended to do for a while, which is package up the subject viewers into their own library and make it available for other contexts. The project app could import it and the subject viewer can be available as part of the selection process in the selection modal. 

I really don't think the classifier should be modified to be a subject explorer module too. I think that will cause many issues for us down the road because there will likely be feature requests for it that I can imagine will be really difficult to accommodate within the classifier context both functionally and in design.

**Reply by snblickhan** [reply]
>which is package up the subject viewers into their own library and make it available for other contexts. The project app could import it and the subject viewer can be available as part of the selection process in the selection modal.

It's really hard to get a sense of what you mean by this since we don't have any design materials to point to, which is why I'm still hesitant to deviate this far from what we'd discussed as a group. 

I'll note that we haven't actually answered the question about whether adding this to the classifier actually requires a major modification, though -- that's what I'm trying to get to the heart of with my questions and user stories. If we don't need to change when a classification starts (re: my questions above about Already Seen behavior), isn't the idea of a 'next' button just a way of baking in the 'refresh' option to the classifier to get the next available subject in order? It's a user experience question, more than anything. I want to make sure it's clear what this process would actually require us to do before we reject it outright.

>Why though? I'm reading through the user stories, and I see this one as part of "Later", which I think maybe is what is driving this?

This was a feature that we originally had categorized as "later," sine we didn't know enough about what the UX would be like to be able to anticipate how much this would be needed. As we got further into the process of building the interface (testing, seeing what things look like, etc), it became clear that this would be a highly desirable feature from the volunteer perspective, and so we decided to consider it for this effort. I'm sorry the User Stories doc didn't get updated, it came up fairly close to the time when Becky was leaving and we didn't get a chance to go over that doc together.

>we can do something that we've intended to do for a while, which is package up the subject viewers into their own library and make it available for other contexts. The project app could import it and the subject viewer can be available as part of the selection process in the selection modal.

I'd be very supportive of exploring this option, but my main concern is that we don't have a designer right now, and this to me reads like something that would really benefit from design thinking. If this process is going to get too complex or balloon up, I'd rather just require volunteers to go back to the indexing page if they don't want to work on a subject. They will likely respond negatively, as that seems cumbersome and would require a lot of additional clicking just to get a new subject, but I'd rather go with something we know isn't ideal (but we know we can build on) than shoehorn in alternate options that we haven't been able to think through in full.

**Reply by srallen** [reply]
I think 'show me a new subject' is something we can easily support and can be something any project can use because we do load up a queue of subjects regardless of project type, but not the 'previous' option. Can we do this iteratively and leave out the previous button? If we find in beta testing that volunteers really want this, then we can then determine how to do this then? 

In addition, I think we can build the option reopen the selection modal from the classifier. What I don't think we should do is make the classifier a read only subject explorer interface. This makes it not a classifier and into something else.

**Reply by snblickhan** [reply]
>Can we do this iteratively and leave out the previous button? If we find in beta testing that volunteers really want this, then we can then determine how to do this then?

I think this seems like a great compromise and would be supportive of exploring this option.

>What I don't think we should do is make the classifier a read only subject explorer interface.

Does this apply to someone who wants to view a completed subject? Related to https://github.com/zooniverse/front-end-monorepo/discussions/2165 -- are you suggesting we shouldn't open those in the classifier at all, but have an alternative view option? Just trying to make sure I'm following the whole thread!

**Reply by srallen** [reply]
> Does this apply to someone who wants to view a completed subject? Related to #2165 -- are you suggesting we shouldn't open those in the classifier at all, but have an alternative view option?

No, we allow anyone to view finished subjects now, however, I wouldn't let that modal permanently block the task area because of the same reasons we allow people to view finished subjects now: citations purposes and educational purposes. We should let them still have the ability to classify even if the classification won't be used.

**Reply by chrislintott** [reply]
I think I'm agreeing with where this conversation is going. My take on this is that the discussion Sam and I are leading on what will replace Talk includes the fully functioning subject/subject set browser that @srallen is worried this might become, and - like Talk - I agree that's separate. 

However, limited ability to move to the 'next' subject is something volunteers want to do **while classifying**, which means while looking at the classifier. I think a separate modal will have a very different effect, and will essentially take people away from classifying into a 'browse' mode (where they really will expect more functionality than is proposed here).

Therefore, I think we should stick to a limited 'next' and 'previous' view in the classifier. This is consistent with how our project partners are expecting this to be implemented. I would also argue that 'previous' should be included too, from the start - if only because if I can go forward, I should be able to go back, else that will be very frustrating and hard to explain.

**Reply by eatyourgreens** [reply]
Once https://github.com/zooniverse/front-end-monorepo/pull/2148 is reviewed and merged, I think 'Previous' would be reasonably straightforward to implement. We can ping Datasette for preceding subject IDs from a subject set, in priority order, then request those IDs from Panoptes to get the serialised subjects, in order, with classification status (Retired, Already Seen etc.) Essentially the same subject selection strategy that's added in https://github.com/zooniverse/front-end-monorepo/pull/2148 but querying for subjects before the current subject, rather than after it.

Is this something that we want to have in place in time for the Royal Botanic Garden project launch? I think we should keep the scope of changes fairly small. Adding buttons to the classifier sounds doable. Extracting the subject viewers out into a library, I'm not sure about in the time available. Might be a good long-term goal, though.

**Comment by snblickhan** [comment]
Current infrastructure supports forward movement _independent of a subject’s status_, e.g. retired/already seen.
- Idea is that we’ll give them the option to ‘scroll’ through the set using Next and Previous page buttons. Note: this will only be available for projects that have enabled sequential classification.

Button that says “New subject” 
- By default, this is the next classifiable subject
- This is something we can implement for projects that haven’t enabled sequential classification, as it essentially replaces the behavior of refreshing the page

Where should these buttons live? 
- Recommendation 1: NOT in the Task area
- Recommendation 2: above the fold

NB: If you’re mid-classification and hit any of these buttons, need a pop up that warns people they are going to lose their work if they leave the page (see https://github.com/zooniverse/front-end-monorepo/issues/2027)

**Reply by srallen** [reply]
Here's the issue for generic version of the suggester modal for finished or already seen subjects: https://github.com/zooniverse/front-end-monorepo/issues/642
